### ISSUE TITLE
Feature: C# bindings and Semantic Kernel integration

### ISSUE THREAD

**Issue by ashvardanian** [unum-cloud/usearch#160]
### Describe what you are looking for

With most of the common programming languages covered, the only major one we don't have bindings for is C#.
Having that will open doors to integrate USearch into Semantic Kernel C# backend, as well as a number of Unity applications.

### Can you contribute to the implementation?

- [X] I can contribute

### Is your feature request specific to a certain interface?

Other bindings

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ashvardanian** [comment]
A couple of recommendations from @dluc:

> 1. I would encourage adopting the default C# code style (settings can be found in the SK repo, e.g. see https://github.com/microsoft/semantic-kernel/blob/main/.editorconfig)
> 2. if possible please make the library .NET Standard 2.0 compatible; that will increase adoption, e.g. across some old apps that haven't migrated to Standard 2.1 or .NET6+

**Comment by dluc** [comment]
Looking forward to this, in particular the disk based index feature. Let me know if I can help! :-)

**Comment by ashvardanian** [comment]
@AleksandrKent has already made some [great progress on the Python side](https://github.com/microsoft/semantic-kernel/compare/main...AleksandrKent:semantic-kernel:usearch) . C# next. 

@dluc, for compatibility and simplicity, we currently put metadata into Parquet files on disk, but integrating UStore down the road would make more sense.

**Comment by dranger003** [comment]
Hi there, I've taken an initial take on some C# bindings, feel free to use any of it if you think this can be useful.
https://github.com/dranger003/usearch-dotnet

**Comment by ashvardanian** [comment]
Oh, wow! Great to have you with us @dranger003! We are also half way through with those, probably makes sense to syn up and merge the progress. @AleksandrKent how is your fork doing?

**Comment by ashvardanian** [comment]
@dranger003 would you like to merge your progress into the main repo under the `csharp/` directory?

**Comment by ashvardanian** [comment]
Thank you, @dranger003! We will continue from there and will release a NuGet package on GitHub.

**Comment by dranger003** [comment]
Hi @AleksandrKent, great work on interop code! I only focused on the low level P/Invoke calls, which have been pulled. Maybe you can sync up and adjust your wrapper? There are few more things I think I need to clean in the interop layer:
- Although `nuint` is the right type for `size_t` it is not widely used at a higher level. I have used `int` for now but I think it should be changed to `ulong` throughout
- I used the same testing code as C99 but I think it's missing a number of test cases, and having some benchmarks would be nice
- As you pointed out, there is currently no support other than f32/f64

**Comment by ashvardanian** [comment]
@dranger003 and @AleksandrKent, is there support for `f16` datatypes in C#? Assuming this may be used with Unity when developing mobile games, one can expect most Arm v8+ CPUs to have half-precision support.

**Comment by dranger003** [comment]
@ashvardanian Yes, I sent another pull request to add `f16` support.

**Comment by AleksandrKent** [comment]
Hi @dranger003 and @ashvardanian,

I've got a few thoughts.

We're using dynamic libraries for C# bindings. What if we added these to the NuGet package? There's guidance in [Microsoft's docs](https://learn.microsoft.com/en-us/nuget/create-packages/supporting-multiple-target-frameworks#architecture-specific-folders). This might lead us to move libusearch build scripts from .csproj to GitHub Actions.

My recent work on C# bindings isn't aligning with our feature branch. I thought of creating a separate C# interface for lib, but maybe the current C lib interface is good for NuGet. @dranger003, can you wrap up your work as a NuGet package?

When we start Semantic Kernel integration, I might be more useful. Should we open a new issue for it once NuGet is ready?

Keen to hear your thoughts.

**Comment by AleksandrKent** [comment]
Hi @dranger003 and @ashvardanian,

I've been reviewing the recent progress on the C# bindings, and I have some thoughts and suggestions to discuss:

1) **Type Casting in `usearch_add`:** Instead of casting to object and then to the proper float type, would it be more efficient to create separate methods for each type and utilize overloading? Generics with if-else statements might not be the best practice for our case.

2) **Error Handling in `usearch_get`:** The error returned from `usearch_dimensions` is lost. To address this:
   2.1) We could separate native calls with the DLLImport attribute into a `NativeMethods` class, and wrappers into another class, as I started in my code. This would create a 1-1 match with C lib functions and handle errors with `Marshal.PtrToStringAnsi(error)` on the public API side.
   2.2) This approach would also eliminate the need for `EntryPoint = "usearch_contains"`, aligning with Microsoft's guidelines on 1-1 namings.
   2.3) We might consider making the `Interop` class internal.

3) **Platform Detection:** I suggest using a static constructor instead of preprocessor directives for platform detection. This would provide a clear exception, as I encountered errors on Ubuntu 22.04 when trying to build with the dotnet CLI.

4) **Calling Convention:** Should we specify the calling convention as `cdecl`? This could be an issue if the target OS uses `stdcall` while the C lib uses `cdecl` by default.

5) **Building on Ubuntu 22.04:** I had trouble running the code on Ubuntu 22.04 via the dotnet CLI. Perhaps we could add a script with a few lines using the dotnet CLI to ensure compatibility. It is also might be helpful to add a script for manual testing at the local environment with the dotnet CLI.

6) **Testing Framework:** I recommend rewriting the tests to align with xUnit, for clarity.
   6.1) Additionally, @ashvardanian suggested auto-generating vectors instead of reading data from files, as they are becoming obsolete.

7) **.gitignore File:** I believe the .gitignore should remain local to your development environment, or we could add specific paths like `csharp/bin` or `csharp/obj` to the core repo's .gitignore.

8) **Purpose of "*.vcxproj" Files:** Are these files necessary for our project? If they are a result of using MSBuild, would it be better to stick with `dotnet` only?

I'm eager to hear your thoughts on these points, and I'm ready to assist with any refactoring based on our considerations.
